docs/create_your_own_brick.rst

```diff
@@ -0,0 +1,387 @@
+brick class (by opposition to class instances).
+
+Application properties
+""""""""""""""""""""""
+
+In the previous examples, the names of the arguments of the application methods
+were directly provided as arguments of the ``@application`` decorator because
+they were common to all instances of the classes. On the other hand, if these
+names need to be defined differently for particular instances of the class,
+you should use the ``apply.property`` decorator. Let's say that we want to
+name our attribute inputs with the string ``self.fancy_name``, then we should
+write:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     def __init__(self, fancy_name):
+    ...         self.fancy_name = fancy_name
+    ...     @application
+    ...     def apply(self, input)
+    ...         ...
+    ...     @apply.property('inputs')
+    ...     def apply_inputs(self):
+    ...         # Note that you can use any python code to define the name
+    ...         return self.fancy_name
+
+Using application calls
+"""""""""""""""""""""""
+
+If you need to add auxiliary variables to the computation graph
+:class:`.ComputationGraph`, you need to pass ``application_call`` as argument
+of your ``apply`` function, as shown in this example:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application
+    ...     def apply(self, x, application_call):
+    ...         application_call.add_auxiliary_variable(x.mean())
+    ...         return x + 1
+
+You can retrieve these annotated variables as usual with the computational
+graph :class:`.ComputationGraph` and filters :class:`.VariableFilter`.
+
+
+Lazy initialization
+~~~~~~~~~~~~~~~~~~~
+
+.. doctest::
+   :hide:
+
+    >>> from blocks.bricks import Feedforward, Linear
+    >>> from blocks.bricks.base import lazy, application
+    >>> from blocks.initialization import Constant
+
+Instead of forcing the user to provide all the brick attributes as arguments
+to the :meth:`.Brick.__init__` method, you could let him/her specify them
+later, after the creation of the brick. To enable this mechanism,
+called lazy initialization, you need to decorate the constructor with the 
+:func:`.lazy` decorator:
+
+    >>> @lazy(allocation=['attr1', 'attr2']) # doctest: +SKIP
+    ... def __init__(self, attr1, attr1)
+    ...     ...
+
+This allows the user to specify ``attr1`` and ``attr2`` after the creation of 
+the brick. For example, the following ``DoubleSequential`` brick is composed of
+two :class:`.Feedforward` bricks for which you do not need to
+specify the ``input_dim`` of ``brick2`` directly at its creation.
+
+    >>> class DoubleSequential(Feedforward):
+    ...     """Two sequential Feedforward bricks."""
+    ...     def __init__(self, brick1, brick2, **kwargs):
+    ...         super(Feedforward, self).__init__(**kwargs)
+    ...         self.brick1 = brick1
+    ...         self.brick2 = brick2
+    ...         self.input_dim = brick1.input_dim
```

Comment:
That's actually an example of a bad behaviour, because `brick1` might not have `input_dim` at that stage. You can see `Sequence` for a more robust example of implementing `Feedforward` interface: https://github.com/mila-udem/blocks/blob/master/blocks/bricks/__init__.py#L641